pkg/transfer/flex/main.go

```diff
@@ -128,91 +154,385 @@ func (volp *DatasetVolumes) readDatasetOpts(mountPath string) (*datasetOpts, err
+	err := os.Remove(path)
+	return errors.Wrap(err, "failed to delete metadata file")
 }
 
-//Init the flex volume
-func (volp *DatasetVolumes) Init() (Capabilities, error) {
-	return Capabilities{Attach: false}, nil
+//Creates a file with a file system inside of it that can be mounted
+func (volp *DatasetVolumes) createFSInFile(path string, filesystem FileSystem, size int64) error {
+	//Create file with room to contain writable file system
+	f, err := os.Create(path)
+	if err != nil {
+		err = errors.Wrap(err, "failed to create file system file")
+		return err
+	}
+
+	err = f.Truncate(size)
+	if err != nil {
+		err = errors.Wrap(err, "failed to allocate file system size")
+		return err
+	}
+
+	//Build file system within
+	cmd := exec.Command("mkfs", "-t", string(filesystem), path)
+	buf := bytes.NewBuffer(nil)
+	cmd.Stderr = buf
+	err = cmd.Run()
+	if err != nil {
+		err = errors.Wrap(errors.New(strings.TrimSpace(buf.String())), "failed to execute mkfs command")
+		return err
+	}
+
+	return nil
 }
 
-//Mount the flex voume, path: '/var/lib/kubelet/pods/c911e5f7-0392-11e8-8237-32f9813bbd5a/volumes/foo~cifs/input', opts: &main.MountOptions{FSType:"", PodName:"imagemagick", PodNamespace:"default", PodUID:"c911e5f7-0392-11e8-8237-32f9813bbd5a", PVOrVolumeName:"input", ReadWrite:"rw", ServiceAccountName:"default"}
-func (volp *DatasetVolumes) Mount(mountPath string, opts MountOptions) error {
-	dsopts, err := volp.writeDatasetOpts(mountPath, opts)
+//Clean up file system in file
+func (volp *DatasetVolumes) destroyFSInFile(path string) error {
+	err := os.RemoveAll(path)
+	if err != nil {
+		err = errors.Wrap(err, "failed to delete fs-in-file file")
+	}
+
+	return err
+}
+
+//Make specified input available at given path (input may be nil)
+func (volp *DatasetVolumes) provisionInput(path string, input *transfer.Ref) error {
+	//Create directory at path in case it doesn't exist yet
+	err := os.MkdirAll(path, DirectoryPermissions)
 	if err != nil {
-		return fmt.Errorf("failed to write volume database: %v", err)
+		return errors.Wrap(err, "failed to create input directory")
 	}
 
-	if dsopts.Input == nil {
-		return nil //no input for volume
+	//Abort if there is nothing to download to it
+	if input == nil {
+		return nil
 	}
 
+	//Download input to it
 	var trans transfer.Transfer
 	if trans, err = transfer.NewS3(&transfer.S3Conf{
-		Bucket: dsopts.Input.Bucket,
+		Bucket: input.Bucket,
 	}); err != nil {
+		return errors.Wrap(err, "failed to set up S3 transfer")
+	}
+
+	ref := &transfer.Ref{
+		Bucket: input.Bucket,
+		Key:    input.Key,
+	}
+
+	err = trans.Download(context.Background(), ref, path)
+	if err != nil {
+		return errors.Wrap(err, "failed to download data from S3")
+	}
+
+	return nil
+}
+
+//Clean up input data
+func (volp *DatasetVolumes) destroyInput(path string) error {
+	err := os.RemoveAll(path)
+	if err != nil {
+		err = errors.Wrap(err, "failed to destroy input directory")
+	}
+
+	return err
```

Comment:
Why not use what you did on https://github.com/nerdalize/nerd/pull/277/files#diff-08539e85d2f6b9f2cca066cf57f3b229R165 ?  It avoids you some checks. Or you can use the traditional way,
```
	if err != nil {
		return errors.Wrap(err, "failed to destroy input directory")
	}

	return nil
```